File: week1/intro.cpp
class Foo {
public:
    mutex m;
    condition_variable cv;
    int turn = 1;
    
    Foo() {
    }

    void printn(function<void()>& printFunc, int n) {
        {
            unique_lock lk(m);
            cv.wait(lk, [this,n](){return turn == n;});
        }
        printFunc();
        {
            unique_lock lk(m);
            turn++;
        }
        cv.notify_all();
    }

    void first(function<void()> printFirst) {
        printn(printFirst,1);
    }

    void second(function<void()> printSecond) {
        printn(printSecond,2);
    }

    void third(function<void()> printThird) {
        printn(printThird,3);
    }
}

File: week1/intro.java
class Foo {
    int order = 1; 
    public Foo() {
        
    }

    public void printn(Runnable printFunc, int n) throws InterruptedException {
        synchronized (this) {
            while(order != n) {
                wait();
            }
        }
        printFunc.run();
        synchronized (this) {
            order++;
            notifyAll();
        }
    }

    public void first(Runnable printFirst) throws InterruptedException {
        printn(printFirst, 1);
    }

    public void second(Runnable printSecond) throws InterruptedException {
        printn(printSecond, 2);
    }

    public void third(Runnable printThird) throws InterruptedException {
        printn(printThird, 3);
    }
